docs/source/miscellaneous/doc/adr/0014-cassandra-targeted-sweep.md

```diff
@@ -0,0 +1,283 @@
+     the maximum exclusive timestamp.
+  2. If the latter occurs, we are guaranteed that there are no entries in *sweepableCells* for the shard and strategy and any of the
+     timestamps in the specified range.
+  3. Otherwise, we now have a fine timestamp partition that is effectively a reference to a row of SweepableCells that is expected
+     to contain at least one cell (this may not be true in degenerate cases where a thread writing or cleaning the queue dies during
+     the process). We now read cells from that row, with start timestamp greater than the last swept timestamp, referring to
+     dedicated rows as necessary, until we either finish the row, exceed the *maxTs*, or we have read 100_000 entries, plus any
+     additional entries to ensure we have read all the entries with the same start timestamp as the latest one.
+
+#### Cleaning the Sweep Queue
+
+Once an entire non-dedicated row of *sweepableCells* or a row of *sweepableTimestamps* are not needed anymore, we remove them as
+follows.
+
+1. Given a shard, strategy, and fine timestamp partition delete all entries from *sweepableCells* for the corresponding row. Note
+   that in Cassandra, an entire row can be deleted at once using a single tombstone.
+    - First, we read through the non-dedicated row to find all the references to dedicated rows and delete them.
+    - Then, delete the the non-dedicated row.
+2. Then, if necessary, given a coarse timestamp partition, also delete the row defined by the shard, strategy, and the coarse
+   timestamp partition in *sweepableTimestamps*.
+
+### Targeted Sweep
+
+Targeted sweep reads the write metadata from the sweep queue instead of sequentially scanning tables to find historic versions.
+After an entry is read from the sweep queue, we check the commit timestamp of the transaction that performed the write.
+
+- If both the start timestamp and the commit timestamp are lower than the sweep timestamp, we can use a single ranged tombstone
+to delete all prior versions of the cell. Note that the Cassandra implementation of the `deleteAllTimestamps` method that writes
+this ranged tombstone does not require reading any information from the KVS and therefore provides a substantial improvement in
+comparison to legacy sweep that needs to find all the previous timestamps so they can be deleted one by one.
+
+- If the commit timestamp is greater than the sweep timestamp, targeted sweep must wait until the sweep timestamp increases
+enough so that the entry can be processed. This is generally not an issue since it is only likely to happen when targeted
+sweep is processing writes that were written within the last hour.
+
+#### Targeted Sweep Implementation
+
+Targeted Sweep has a number of background threads (as controlled the install config) continuously cycling through the shards for
+each strategy. To find the next shard to sweep, the thread requests a TimeLock lock for the shard and strategy. If successful, it
```

Comment:
I know what's going on here, but might be worth making things more explicit that we have a bunch of threads running _for each strategy_, that iterate through the shards. I'm not sure what a good way to justify that is (as opposed to say ordering ShardAndStrategy by shard and then arbitrarily saying CONSERVATIVE < THOROUGH, say) though we might have discussed it